Lib/importlib/_bootstrap_external.py

```diff
@@ -310,7 +320,18 @@ def cache_from_source(path, debug_override=None, *, optimization=None):
         if not optimization.isalnum():
             raise ValueError('{!r} is not alphanumeric'.format(optimization))
         almost_filename = '{}.{}{}'.format(almost_filename, _OPT, optimization)
-    return _path_join(head, _PYCACHE, almost_filename + BYTECODE_SUFFIXES[0])
+    filename = almost_filename + BYTECODE_SUFFIXES[0]
+    if bytecode_path is not None:
+        if not _path_isabs(head):
+            head = _path_join(_os.getcwd(), head)
+        if head[1] == ':':
+            head = head[2:]
+        return _path_join(
+            bytecode_path,
+            head.strip(path_separators),
```

Comment:
It would I think, but we can't use `os.path.join` here in `_bootstrap_external`, we don't have access to it. That's why we're using the substitute `_path_join` defined here in the module, which isn't nearly as sophisticated.

Reply:
We would need the `lstrip` part of this even if we were using `os.path.join`, or else it would erase the initial `bytecode_path`. Part of what we're doing here is turning `head` from an absolute path back into a relative path we can join to `bytecode_path`.